Cython/Compiler/ParseTreeTransforms.py

```diff
@@ -3153,7 +3153,8 @@ def _inject_class(self, node):
 
         if class_scope.is_py_class_scope:
             def_node.has_class_reference = True
-            return ExprNodes.NameNode(node.pos, name=class_node.target.name)
+            self_node = ExprNodes.NameNode(pos=node.pos, name=u'self')
+            return ExprNodes.AttributeNode(pos=node.pos, obj=self_node, attribute=EncodedString(u'__class__'))
```

Comment:
No, `self.__class__` is not correct since it's not compile time constant. This should refer to the class that defines the method, not the one on which the method is called. The reference should come from `ClassCellNode` (or something similar).